### ISSUE TITLE
Too eager publishDiagnostics for .hpp even after receiving didChange for it

### ISSUE THREAD

**Issue by joaotavora** [clangd/clangd#1088]
When working with a `foo.hpp` header file and `foo.cpp` implementation file and renaming a symbol, clangd seems to make the renamings correctly but then mistankenly reports an error, since it seems to have used a stale version of the `foo.hpp` whose renaming it itself proposed and which was confirmed.

```c++
// foo.hpp
struct Foo{
  int _foo;
  auto foo() {return _foo;}
};
```
```c++
// foo.cpp
#include "foo.hpp"
int main() { return Foo()._foo; }
```

The logs I attach show the almost-correct renaming of member variable `_foo` to `_bar`.  `clangd` gets the two `didChange` notifs, and then produces two `publishDiagnostics` notifs.  But while one of them is fine, the one for `foo.cpp` isn't.  It's like it was obtained using the old version of the `foo.hpp` file.

Saving the `foo.hpp` file (makes a `didSave`) fixes the situation.

Note that the stderr logs are intermingled with logs from my LSP client, Eglot.  Remove every line not prefixed `[stderr]` if that bothers you.

[logs.txt](https://github.com/clangd/clangd/files/8380045/coiso.txt)

**System information**

Output of `clangd --version`:
```
clangd version 13.0.1
Features: linux
Platform: x86_64-pc-linux-gnu
```

Editor/LSP plugin:  Emacs ([eglot](https://github.com/joaotavora/eglot)

Operating system: ArchLinux

**Comment by sam-mccall** [comment]
Yeah, this is a really unfortunate consequence of the way we see version contents.
When parsing any particular file, we read the headers it includes from disk, even if that file is open with different contents ("dirty").

So in this case, we apply the rename to several open buffers, but the changes to foo.hpp aren't visible to foo.cpp until you save the header. The diagnostics are "correct" in the sense that they're the same ones you'd get if you applied the edits by hand.

The basic reason for this is performance: any change to any headers included by foo.cpp requires a full reparse, which is very expensive. By only making header changes visible on save, we reduce the frequency of this.

[clangd 14](https://github.com/clangd/clangd/releases/tag/14.0.0) has the flag `--use-dirty-headers` to change this behavior, please try this if you can and let us know a) if it solves the problem b) if it feels slower in general.

**Comment by sam-mccall** [comment]
CC @njames93

**Comment by joaotavora** [comment]
> The basic reason for this is performance: any change to any headers included by foo.cpp requires a full reparse, which is very expensive. By only making header changes visible on save, we reduce the frequency of this.

Makes sense. But don't you think that in some situations, such as having been the recent target of a successful rename, the "dirty header" should be automatically used? What I suppose happens is that users (or even clients) will save immediately after rename to avoid this, which will amount to the same full expensive full reparse in the end (even slightly more expensive, since it'll probably come on top of the cheap one). 

So, I haven't tried `--use-dirty-headers` yet.  If it ends up being too slow, I think a heuristic that looks at the most recent rename that targeted `foo.cpp`, finds out which headers were included in it makes sense and uses the dirty version of those makes sense.

If it's not too slow, then my problem will be solved :-)

**Comment by njames93** [comment]
There is a little nit I have found with multi-file edits. Sometimes the header file edit is registered after a source file edit, so when we rebuild the AST for the source file it will still have the old header file. This can give erroneous errors, but the go away after the file is rebuild again(following say an edit to the file)
